xarray/core/dataarray.py

```diff
@@ -3430,6 +3432,225 @@ def pad(
+        if not self.ndim:
+            ValueError("This function does not apply for scalars")
+
+        if dim is not None:
+            pass  # Use the dim if available
+        elif axis is not None:
+            dim = self.dims[axis]
+        elif self.ndim == 1:
+            # it is okay to guess the dim if there is only 1
+            dim = self.dims[0]
+        else:
+            # The dim is not specified and ambiguous.  Don't guess.
+            raise ValueError(
+                "Must supply either 'dim' or 'axis' argument "
+                "for multidimensional arrays"
+            )
+
+        if dim in self.coords:
+            pass  # This is okay
+        elif axis is not None:
+            raise IndexError(f'Axis "{axis}" does not have coordinates')
+        else:
+            raise KeyError(f'Dimension "{dim}" does not have coordinates')
+
+        # These are dtypes with NaN values argmin and argmax can handle
+        na_dtypes = "cf0"
+
+        if skipna or (skipna is None and self.dtype.kind in na_dtypes):
+            # Need to skip NaN values since argmin and argmax can't handle them
+            allna = self.isnull().all(dim)
+            array = self.where(~allna, 0)
+            hasna = allna.any()
+        else:
+            array = self
+            allna = None
+            hasna = False
+
+        # If promote is None we only promote if there are NaN values.
+        if promote is None:
+            promote = hasna
+
+        if not promote and hasna and array.coords[dim].dtype.kind not in na_dtypes:
+            raise TypeError(
+                "NaN values present for NaN-incompatible dtype and Promote=False"
+            )
+
+        # This will run argmin or argmax.
+        indx = getattr(array, func)(
+            dim=dim, axis=None, keep_attrs=False, skipna=skipna, **kwargs
+        )
+
+        # Get the coordinate we want.
+        coordarray = array[dim]
+
+        # Handle dask arrays.
+        if isinstance(array, dask_array_type):
+            res = dask_array.map_blocks(coordarray, indx, dtype=indx.dtype)
+        else:
+            res = coordarray[
+                indx,
+            ]
+
+        # Promote to a dtype that can handle NaN values if needed.
+        newdtype, fill_value = dtypes.maybe_promote(res.dtype)
+        if promote and newdtype != res.dtype:
+            res = res.astype(newdtype)
+
+        # Put the NaN values back in after removing them, if necessary.
+        if hasna and allna is not None:
+            res = res.where(~allna, fill_value)
+
+        # The dim is gone but we need to remove the corresponding coordinate.
+        del res.coords[dim]
+
+        # Put the attrs back in if needed
+        if keep_attrs:
+            res.attrs = self.attrs
+
+        return res
+
+    def idxmin(
+        self,
+        dim: Optional[Hashable] = None,
+        axis: Optional[int] = None,
+        skipna: Optional[bool] = None,
+        promote: Optional[bool] = None,
```

Comment:
Nit: no need for `Optional` if the default is `None`

Reply:
Fixed, will submit a new pull request with all changes

Reply:
you could also push to this PR? I don't think you need to worry about the commit history, we always do a squash-merge so in master this whole PR ends up as a single commit.

Reply:
Yes fine either way; no need to get everything in here, but no need to do a separate PR if you've already done the code!

Reply:
Sorry, I misspoke.  I meant new commit, not new pull request.  The new commit has already been pushed to this branch, please take a look.